src/Microsoft.AspNetCore.Routing/Template/TemplateBinder.cs

```diff
@@ -342,7 +342,14 @@ private static bool IsRoutePartNonEmpty(object routePart)
                 var hasAmbientValue = ambientValues != null &&
                     ambientValues.TryGetValue(keyName, out ambientValue);
 
-                if (hasAmbientValue && !RoutePartsEqual(ambientValue, explicitValue))
+                // This indicates an explicit value was provided whose key does not exist in ambient values
+                // Example: Link from controller-action to a page or vice versa.
+                if (!hasAmbientValue)
+                {
+                    return false;
+                }
```

Comment:
I'm not sure that this is right. MVC should handle this case by making sure that all of the endpoints have appropriate `null` values. Basically, I would expect the set of required keys to be `controller, action, area, page` for all actions, for an application that uses areas, controllers, and pages.

Reply:
Basically, I think your test setup is wrong